File: InteractiveChartView.m
//
//  InteractiveChartView.m
//  ShralpTide2
//
//  Created by Michael Parlee on 11/2/13.
//
//

#import "InteractiveChartView.h"

#define CURSOR_TOP_GAP 40
#define CURSOR_LABEL_WIDTH 3

@interface InteractiveChartView ()

@property (nonatomic, strong) NSMutableDictionary *times;

@end

@implementation InteractiveChartView

- (id)initWithCoder:(NSCoder *)aDecoder
{
    self = [super initWithCoder:aDecoder];
    if (self) {
		self.times = [[NSMutableDictionary alloc] init];
        self.cursorView = [[CursorView alloc] initWithFrame:CGRectMake(0, CURSOR_TOP_GAP, CURSOR_LABEL_WIDTH, self.frame.size.width - CURSOR_TOP_GAP)];
    }
    return self;
}

- (int)currentTimeInMinutes {
	// The following shows the current time on the tide chart.  Need to make sure that it only shows on
	// the current day!
	NSDate *datestamp = [NSDate date];
	NSDate *midnight = [self midnight];
    
	if ([midnight compare:[self.datasource day]] == NSOrderedSame) {
		return ([datestamp timeIntervalSince1970] - [midnight timeIntervalSince1970]) / SECONDS_PER_MINUTE;
	} else {
		return -1;
	}
}

- (int)currentTimeOnChart
{
    return [self currentTimeInMinutes] * self.frame.size.width / MINUTES_PER_HOUR * self.hoursToPlot;
}

-(void)showTideForPoint:(CGPoint) point {
    NSDate *dateTime = [self dateTimeFromMinutes:point.x];
    [self.delegate displayHeight: point.y
                          atTime: dateTime
                  withUnitString: [[self.datasource tideDataToChart] unitShort]];
}

- (NSDate*)dateTimeFromMinutes:(int)minutesSinceMidnight
{
	NSString* key = [NSString stringWithFormat:@"%d",minutesSinceMidnight];
	if ((self.times)[key] != nil) {
		return (self.times)[key];
	} else {
		int hours = minutesSinceMidnight / 60;
		int minutes = minutesSinceMidnight % 60;
		
		NSCalendar *gregorian = [NSCalendar currentCalendar];
		NSDateComponents *components = [[NSDateComponents alloc] init];
		[components setHour:hours];
		[components setMinute:minutes];
		
		NSDate *time = [gregorian dateByAddingComponents:components toDate:[self.datasource day] options:0];
        return time;
	}
}

#pragma mark HandleTouch
- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {
//    DLog(@"Touches began on chartview!");
    // We only support single touches, so anyObject retrieves just that touch from touches
    UITouch *touch = [touches anyObject];
    
    // Animate the first touch
    CGPoint touchPoint = [touch locationInView:self];
    CGPoint dataPoint = [[self.datasource tideDataToChart] nearestDataPointForTime:[self timeInMinutes:touchPoint.x]];
	CGPoint movePoint = CGPointMake(touchPoint.x, self.frame.size.height / 2 + CURSOR_TOP_GAP);
	
	if (self.cursorView.superview == nil) {
		[self addSubview:self.cursorView];
	}
	
    [self animateFirstTouchAtPoint:movePoint];
	[self showTideForPoint: [[self.datasource tideDataToChart] nearestDataPointForTime: [self timeInMinutes:touchPoint.x]]];
}


- (void)touchesMoved:(NSSet *)touches withEvent:(UIEvent *)event {
    UITouch *touch = [touches anyObject];
    CGPoint touchPoint = [touch locationInView:self];
    CGPoint dataPoint = [[self.datasource tideDataToChart] nearestDataPointForTime:[self timeInMinutes:touchPoint.x]];
    CGPoint movePoint = CGPointMake(touchPoint.x, self.frame.size.height / 2 + CURSOR_TOP_GAP);
    self.cursorView.center = movePoint;
    [self showTideForPoint: dataPoint];
}

- (int)timeInMinutes:(CGFloat)xPosition
{
    return (xPosition / self.xratio);
}


- (void)touchesEnded:(NSSet *)touches withEvent:(UIEvent *)event {
//    DLog(@"Touches ended.");
    self.userInteractionEnabled = NO;
    [self animateCursorViewToCurrentTime];

	// if not the current day hide the cursor and tide details
	if (self.cursorView.center.x <= 0.0) {
		[self.cursorView removeFromSuperview];
	}
}


- (void)touchesCancelled:(NSSet *)touches withEvent:(UIEvent *)event {
//    DLog(@"Touches cancelled");
    /*
     To impose as little impact on the device as possible, simply set the cursor view's center and transformation to the original values.
     */
    self.cursorView.center = self.center;
    self.cursorView.transform = CGAffineTransformIdentity;
}

- (void)animateFirstTouchAtPoint:(CGPoint)touchPoint {
    
#define MOVE_ANIMATION_DURATION_SECONDS 0.15
    
    NSValue *touchPointValue = [NSValue valueWithCGPoint:touchPoint];
    [UIView beginAnimations:nil context:(__bridge void *)(touchPointValue)];
    [UIView setAnimationDuration:MOVE_ANIMATION_DURATION_SECONDS];
    CGPoint movePoint = CGPointMake(touchPoint.x, self.frame.size.height / 2 + CURSOR_TOP_GAP);
    self.cursorView.center = movePoint;
    [UIView commitAnimations];
}

- (void)animateCursorViewToCurrentTime {
    if (self.cursorView.superview == nil) {
		[self addSubview:self.cursorView];
	}
    
    // Bounces the placard back to the center
	
    CALayer *welcomeLayer = self.cursorView.layer;
    
    // Create a keyframe animation to follow a path back to the center
    CAKeyframeAnimation *bounceAnimation = [CAKeyframeAnimation animationWithKeyPath:@"position"];
    bounceAnimation.removedOnCompletion = NO;
    
    CGFloat animationDuration = 0.5;
	
    // Create the path for the bounces
    CGMutablePathRef thePath = CGPathCreateMutable();
    
    CGFloat midX = [self currentTimeInMinutes] * self.xratio;
    CGPoint dataPoint = [[self.datasource tideDataToChart] nearestDataPointForTime:[self timeInMinutes:midX]];
    CGFloat midY = self.frame.size.height / 2 + CURSOR_TOP_GAP;
    CGFloat originalOffsetX = self.cursorView.center.x - midX;
    CGFloat originalOffsetY = self.cursorView.center.y - midY;
    CGFloat offsetDivider = 10.0;
    
    BOOL stopBouncing = NO;
    
    // Start the path at the cursors's current location
    CGPathMoveToPoint(thePath, NULL, self.cursorView.center.x, self.cursorView.center.y);
    CGPathAddLineToPoint(thePath, NULL, midX, midY);
    
    // Add to the bounce path in decreasing excursions from the center
    while (stopBouncing != YES) {
        CGPathAddLineToPoint(thePath, NULL, midX + originalOffsetX/offsetDivider, midY + originalOffsetY/offsetDivider);
        CGPathAddLineToPoint(thePath, NULL, midX, midY);
		
        offsetDivider += 10;
        animationDuration += 1/offsetDivider;
        if ((fabs(originalOffsetX/offsetDivider) < 6)) {
            stopBouncing = YES;
        }
    }
    
    bounceAnimation.path = thePath;
    bounceAnimation.duration = animationDuration;
	
    
    // Create a basic animation to restore the size of the placard
    CABasicAnimation *transformAnimation = [CABasicAnimation animationWithKeyPath:@"transform"];
    transformAnimation.removedOnCompletion = YES;
    transformAnimation.duration = animationDuration;
    transformAnimation.toValue = [NSValue valueWithCATransform3D:CATransform3DIdentity];
    
    
    // Create an animation group to combine the keyframe and basic animations
    CAAnimationGroup *theGroup = [CAAnimationGroup animation];
    
    // Set self as the delegate to allow for a callback to reenable user interaction
    theGroup.delegate = self;
    theGroup.duration = animationDuration;
    theGroup.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseIn];
    
    theGroup.animations = @[bounceAnimation, transformAnimation];
    
    
    // Add the animation group to the layer
    [welcomeLayer addAnimation:theGroup forKey:@"animatePlacardViewToCenter"];
    
    // Set the placard view's center and transformation to the original values in preparation for the end of the animation
    self.cursorView.center = CGPointMake(midX, midY);
    self.cursorView.transform = CGAffineTransformIdentity;
    
    CGPathRelease(thePath);
	
	[self showTideForPoint: [self.datasource.tideDataToChart nearestDataPointForTime:[self timeInMinutes:midX]]];
    [self.delegate interactionsEnded];
}


- (void)animationDidStop:(CAAnimation *)theAnimation finished:(BOOL)flag {
    //Animation delegate method called when the animation's finished:
    // restore the transform and reenable user interaction
    //self.cursorView.transform = CGAffineTransformIdentity;
    self.userInteractionEnabled = YES;
}

@end

File: Classes/Views/InteractiveChartView.swift
//
//  InteractiveChartView.swift
//  ShralpTide2
//
//  Created by Michael Parlee on 1/5/19.
//

import Foundation
import UIKit
import QuartzCore
import CoreGraphics

@objc class InteractiveChartView: LabeledChartView, CAAnimationDelegate {
    
    static let CursorTopGap = 40
    static let CursorLabelWidth = 3
    
    @IBOutlet var delegate: InteractiveChartViewDelegate!
    
    var times = Dictionary<String, Date>()
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        cursorView = CursorView(frame: CGRect(x: 0, y: InteractiveChartView.CursorTopGap, width: InteractiveChartView.CursorLabelWidth, height: Int(frame.size.width - CGFloat(InteractiveChartView.CursorTopGap))))
    }
    
    fileprivate func currentTimeInMinutes() -> Int {
        // The following shows the current time on the tide chart.
        // Need to make sure that it only shows on the current day!
        let datestamp = Date()
        
        if midnight() == self.datasource.day {
            return Int(datestamp.timeIntervalSince1970 - self.midnight().timeIntervalSince1970) / InteractiveChartView.SecondsPerMinute
        } else {
            return -1
        }
    }
    
    fileprivate func currentTimeOnChart() -> Int {
        return Int(round(CGFloat(currentTimeInMinutes()) * frame.size.width)) / ChartView.MinutesPerHour * hoursToPlot
    }
    
    func showTide(forPoint point: CGPoint) {
        let dateTime = self.dateTime(fromMinutesSinceMidnight: Int(point.x))
        delegate.displayHeight(point.y, atTime: dateTime, withUnitString: datasource.tideDataToChart.unitShort)
    }
    
    fileprivate func dateTime(fromMinutesSinceMidnight minutes: Int) -> Date {
        let key = String(format: "%d", minutes)
        if self.times[key] != nil {
            return self.times[key]!
        } else {
            let hours = minutes / ChartView.MinutesPerHour
            let minutes = minutes % ChartView.MinutesPerHour
            
            let cal = Calendar.current
            var components = DateComponents()
            components.hour = hours
            components.minute = minutes
            return cal.date(byAdding: components, to: datasource.day)!
        }
    }
    
    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        // We only support single touches, so we retrieve just that touch from touches
        guard let touch = touches.first else {
            return
        }
        
        // Animate first touch
        let touchPoint = touch.location(in: self)
        let movePoint = CGPoint(x: touchPoint.x, y: frame.size.height / 2 + CGFloat(InteractiveChartView.CursorTopGap))
        
        if cursorView.superview == nil {
            addSubview(self.cursorView)
        }
        
        animateFirstTouch(touchPoint: movePoint)
        showTide(forPoint: self.datasource.tideDataToChart.nearestDataPoint(forTime: timeInMinutes(touchPoint.x)))
    }
    
    override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
        guard let touch = touches.first else {
            return
        }
        let touchPoint = touch.location(in: self)
        let dataPoint = datasource.tideDataToChart.nearestDataPoint(forTime: timeInMinutes(touchPoint.x))
        let movePoint = CGPoint(x: touchPoint.x, y: frame.size.height / 2 + CGFloat(InteractiveChartView.CursorTopGap))
        
        cursorView.center = movePoint
        showTide(forPoint: dataPoint)
    }
    
    fileprivate func timeInMinutes(_ xPosition: CGFloat) -> Int {
        return Int(round(xPosition / self.xratio))
    }
    
    override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
        isUserInteractionEnabled = false
        animateCursorViewToCurrentTime()
        
        if cursorView.center.x <= 0.0 {
            cursorView.removeFromSuperview()
        }
    }
    
    override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) {
        cursorView.center = self.center
        cursorView.transform = CGAffineTransform.identity
    }
    
    func animateFirstTouch(touchPoint: CGPoint) {
        let durationSec = 0.15
        UIView.animate(withDuration: durationSec, animations: {
            let movePoint = CGPoint(x: touchPoint.x, y: self.frame.size.height / 2 + CGFloat(InteractiveChartView.CursorTopGap))
            self.cursorView.center = movePoint
        })
    }
    
    override func draw(_ rect: CGRect) {
        super.draw(rect)
        if tide == nil { return }
        animateCursorViewToCurrentTime()
    }
    
    func animateCursorViewToCurrentTime() {
        if cursorView.superview == nil {
            cursorView.frame = CGRect(x: 0, y: 0, width: InteractiveChartView.CursorLabelWidth, height: Int(frame.size.width))
            addSubview(cursorView)
        }
        
        // Bounces the placard back to the center
        let welcomeLayer: CALayer = cursorView.layer
        
        // Create a keyframe animation to follow a path back to the center
        let bounceAnimation = CAKeyframeAnimation(keyPath: "position")
        bounceAnimation.isRemovedOnCompletion = false
        
        var animationDuration:CGFloat = 0.5
        
        // Create the path for the bounces
        let thePath = CGMutablePath()
        
        let midX = CGFloat(currentTimeInMinutes()) * xratio
        let midY = frame.size.height / 2 + CGFloat(InteractiveChartView.CursorTopGap)
        let originalOffsetX = cursorView.center.x - midX
        let originalOffsetY = cursorView.center.y - midY
        
        var offsetDivider: CGFloat = 10.0
        var stopBouncing = false
        
        // Start the path at the cursor's current location
        thePath.move(to: CGPoint(x: cursorView.center.x, y: cursorView.center.y))
        thePath.addLine(to: CGPoint(x: midX, y: midY))
        
        // Add to the bound path in decreasing excursions from the center
        while stopBouncing != true {
            thePath.addLine(to: CGPoint(x: midX + originalOffsetX/offsetDivider, y: midY + originalOffsetY/offsetDivider))
            thePath.addLine(to: CGPoint(x: midX, y: midY))
            
            offsetDivider += 10
            animationDuration += 1/offsetDivider
            if abs(originalOffsetX/offsetDivider) < 6 {
                stopBouncing = true
            }
        }
        
        bounceAnimation.path = thePath
        bounceAnimation.duration = TimeInterval(animationDuration)
        
        // Create a basic animation to restore the size of the placard
        let transformAnimation = CABasicAnimation(keyPath: "transform")
        transformAnimation.isRemovedOnCompletion = true
        transformAnimation.duration = TimeInterval(animationDuration)
        transformAnimation.toValue = NSValue(caTransform3D:CATransform3DIdentity)
        
        // Create an animation group to combine the keyframe and basic animations
        let theGroup = CAAnimationGroup()
        
        // Set self as the delegate to allow for a callback to reenable user interaction
        theGroup.delegate = self
        theGroup.duration = TimeInterval(animationDuration)
        theGroup.timingFunction = CAMediaTimingFunction(name: CAMediaTimingFunctionName.easeIn)
        
        theGroup.animations = [bounceAnimation, transformAnimation]
        
        // Add the animation group to the layer
        welcomeLayer.add(theGroup, forKey: "animatePlacardViewToCenter")
        
        // Set the placard view's center and transformation to the original values in preparation for the end of the animation
        cursorView.center = CGPoint(x: midX, y: midY)
        cursorView.transform = CGAffineTransform.identity
        
        showTide(forPoint: datasource.tideDataToChart.nearestDataPoint(forTime: timeInMinutes(midX)))
        delegate.interactionsEnded()
    }
    
    func animationDidStop(_ anim: CAAnimation, finished flag: Bool) {
        //Animation delegate method called when the animation's finished:
        // restore the transform and reenable user interaction
        self.isUserInteractionEnabled = false
    }
}

